NF: #18 new `bin()` and `downsample()` functions

shawntz opened the pull request:
## summary

This PR introduces new `bin()` and `downsample()` functions raised in #18 for pupil time series data processing, including anti-aliasing filtering for downsampling and averaging for binning. Furthermore, these new functions are integrating into the `eyeris::glassbox()` pipeline.

### `downsample()`

This function downsamples pupil time series data with anti-aliasing filtering and is part of the `eyeris::glassbox()` preprocessing pipeline.

### `bin()`

This function bins time series pupil data by averaging within time bins and is intended for use within the `eyeris::glassbox()` preprocessing pipeline.

shawntz committed:
TEST: #18 add tests for new `downsample()` and `bin()` functions

Introduces a new test file to verify the behavior of the downsample and bin functions, including their integration within the glassbox workflow. Tests check for correct sample rates, reduced row counts, and absence of NA values.

shawntz committed:
TEST: add GitHub Actions workflow for testthat

Introduces a CI workflow to run testthat tests on push and pull requests to main and master branches. The workflow sets up R, installs dependencies, runs tests with devtools, and outputs test results.

shawntz committed:
CHORE: add devtools to Suggests in DESCRIPTION

The devtools package was added to the Suggests field to facilitate development and testing workflows.

shawntz committed:
DOC: #18 add testthat badge and update README example output

Added a testthat GitHub Actions badge to both README.Rmd and README.md to reflect CI status. Updated the example output in README.md to include skipped steps for downsample() and bin() in the eyeris_preproc example.

shawntz committed:
NF: #18 add new `binning` and `downsampling` funcs to `glassbox` pipeline

Introduces new `bin` and `downsample` functions for time series data processing, including anti-aliasing filtering for downsampling and averaging for binning. Integrates both steps into the glassbox pipeline, updates the handler to support their data structure changes, and documents the new functions in the NAMESPACE and pkgdown config. Default parameter handling and warnings are also added for incorrect usage.

shawntz committed:
DOC: #18 add documentation for new `downsample()` function

Introduces the man/downsample.Rd file with detailed documentation for the downsample function, including usage, arguments, return value, and examples. This function downsamples time series data with anti-aliasing filtering and is part of the glassbox preprocessing pipeline.

shawntz committed:
DOC: #18 add documentation for new `bin()` function

Introduces the man/bin.Rd file with documentation for the `bin` function, describing its usage, arguments, value, and examples. This function bins time series pupillometry data by averaging within time bins and is intended for use within the glassbox preprocessing pipeline.

shawntz commented:
Note: there are still some implementational issues with bridging the functionality to the `glassbox()` that I'm actively working through; however, the main math/function logic for `bin()` and `downsample()` should be good for review now @mh105

mh105 commented on R/pipeline-bin.R:
```diff
@@ -0,0 +1,156 @@
+  # create bin centers (1/2X, 3/2X, 5/2X, ...)
+  bin_duration <- 1 / bins_per_second
+  max_time <- max(time_secs_inferred, na.rm = TRUE)
+  bin_centers <- seq(bin_duration / 2, max_time, by = bin_duration)
```
I'm not sure using `seq(bin_duration / 2, max_time)` is the most flexible/safest here. Could we inherit the time vector `time_secs_inferred` to create the new time vector? I.e., it will be something like `seq(time_secs_inferred[0] + bin/duration / 2, max_time)`. Otherwise if time_secs_inferred` doesn't start from 0, this vector will be of wrong length, at least in my mind without thinking too carefully.

shawntz replied:
Thanks for catching this, Alex! 🫡

mh105 commented on R/pipeline-bin.R:
```diff
@@ -0,0 +1,156 @@
+
+  # create bin centers (1/2X, 3/2X, 5/2X, ...)
+  bin_duration <- 1 / bins_per_second
+  max_time <- max(time_secs_inferred, na.rm = TRUE)
```
Do we check anywhere that `x[[time_col]]` is sorted throughout `eyeris`? I'm a little worried about using `max_time <- max(time_secs_inferred, na.rm = TRUE)` in case there are some wonky large values due to `EDF` file errors or something. It might be best to validate that it is a monotonically increasing vector somewhere?

shawntz replied:
Thanks for pointing this out -- I've added checks throughout the package! 🚀

mh105 commented on R/pipeline-downsample.R:
```diff
@@ -0,0 +1,159 @@
+
+  # design anti-aliasing filter
+  fs_nq <- current_fs / 2
+  foo <- gsignal::buttord(wp / fs_nq, ws / fs_nq, 1, 35)  # rp = 1, rs = 35
```
Maybe allow users to pass in `rp` and `rs` as well just like in `lpfilt()`?

shawntz replied:
Done! 💯

mh105 reviewed (changes requested):
Thanks @shawntz ! Just a few small comments and otherwise feel free to merge.

shawntz committed:
FF: #18 improve time axis handling in `plot.eyeris`

Refactors time axis calculations to use consistent units (seconds and milliseconds) and ensures correct scaling for plotting pupil data. Adds more robust handling of sampling rate, improves random epoch selection, and clarifies axis labeling. Also adds informative alerts for sampling rate and improves placeholder data handling for missing samples.

shawntz committed:
FF: #18 fix NA vertical line positions in robust_plot

Vertical lines indicating NAs are now drawn at the correct x values if x is provided, improving plot accuracy in robust_plot.

shawntz committed:
NF: #18 add time_scaled and decimated.sample.rate fields

Introduces a new 'time_scaled' column alongside 'time_secs' in the timeseries normalization process, and adds a 'decimated.sample.rate' field (initialized as NA_integer_) to the output list in load_asc. These changes prepare the data structure for future processing steps that may require scaled time and decimated sample rate information.

shawntz committed:
RF: #18 bin/downsample handling in `pipeline_handler()`

Consolidates bin and downsample operations into the main processing loop, updating the data structure and relevant metadata in place. Adds a guard to ensure 'time_scaled' mirrors 'time_secs' when no downsampling or binning occurs. Removes redundant code and improves clarity in updating the 'latest' pointer.

shawntz committed:
RF: #18 `downsample()` function and improve modularity

Refactors the `downsample()` and internal `downsample_pupil()` functions for improved modularity and clarity. Moves pipeline handling logic out of downsample and into downsample_pupil, simplifies argument passing, and ensures decimated sample rate is tracked. Updates documentation and example usage to reflect changes.

shawntz committed:
RF: #18 and improve `bin()` and internal `bin_pupil()` functions for clarity

Enhanced the bin_pupil function with better error handling, clearer logic, and more robust binning for both numeric and non-numeric columns. The function now checks for valid input, uses a helper for binning, and provides progress feedback when processing multiple columns. Documentation and example usage were also updated for clarity.

shawntz committed:
RF: #18 improve `glassbox()` pipeline parameter handling and validation

Refines parameter handling in the glassbox function, including explicit handling of the 'method' parameter for binning, improved default parameter assignment, and stricter validation to prevent simultaneous use of downsampling and binning. Also enhances warning and info messages, and ensures preservation of decimated.sample.rate and latest pointers from processed blocks.

shawntz committed:
TEST: #18 update tests for `downsample()` and `bin()` output structure

Refactor tests to match updated output structure where decimated and binned sample rates are now top-level fields instead of nested in 'info'. Also update NA checks to target specific columns, and remove unnecessary expect_silent wrappers.

shawntz committed:
FF: #18 use dynamic time range for preview_window in plot example

Updated the README examples to calculate start_time and end_time dynamically from the data for the preview_window argument, instead of hardcoding the range. This makes the example more robust to changes in the data.

shawntz committed:
DOC: #18 update documentation for pupil time series functions

Refine titles and examples in bin and downsample documentation to specify pupil time series data. Update glassbox documentation to use dynamic preview window values. Refresh example plot images to reflect these changes.

shawntz committed:
FF: #205 anchor bin centers to start of time vector in `bin_pupil()`

Updated the calculation of bin centers in the bin_pupil function to start from the minimum time value, ensuring bins are correctly aligned with the data's time range.

Co-Authored-By: Mingjian He <49259768+mh105@users.noreply.github.com>

shawntz committed:
TEST: #18 wrap `glassbox()` tests in `expect_silent` and add `verbose=FALSE`

The glassbox tests for downsampling and binning are now wrapped in expect_silent to suppress output during testing. The verbose=FALSE argument is also added to further reduce unnecessary console output + to test whether all verbose statements are working as expected here.

shawntz committed:
DOC: #18 update `glassbox()` example to clarify method argument

The example in the documentation now explicitly includes the 'method' argument within the 'bin' list for clarity, reflecting the default 'mean' method usage.

shawntz committed:
Update README.md

shawntz committed:
FF: #18 remove redundant info alert in `pipeline_handler()`

Eliminated an informational alert about decimating the sampling rate during bin/downsample operations to reduce unnecessary console output.

shawntz committed:
ENH: #18 add filter parameter options to `downsample()` function

The `downsample()` function now accepts 'rp' (passband ripple) and 'rs' (stopband attenuation) as parameters, allowing users to customize the anti-aliasing filter. Documentation and examples have been updated to reflect these changes.

Co-Authored-By: Mingjian He <49259768+mh105@users.noreply.github.com>

shawntz committed:
DOC: #18 update `downsample()` docs for new filter parameters

Added documentation for new 'rp' (passband ripple) and 'rs' (stopband attenuation) parameters in the downsample function. Updated usage, arguments, and examples to reflect these changes and clarify default values.

Co-Authored-By: Mingjian He <49259768+mh105@users.noreply.github.com>

shawntz committed:
ENH: #18 add default downsample params and pass rp, rs to function

Set default values for downsample parameters 'plot_freqz', 'rp', and 'rs' if not provided. Pass 'rp' and 'rs' to the `eyeris::downsample()` function to allow more control over the downsampling process.

Co-Authored-By: Mingjian He <49259768+mh105@users.noreply.github.com>

shawntz committed:
ENH: #18 add verbose info for downsample and bin steps

Adds a CLI success alert when decimating the sampling rate during the 'downsample' or 'bin' steps if verbose mode is enabled. This provides clearer feedback to users about changes in sampling rate during processing.

shawntz committed:
ENH: #18 add verbose flag to plot.eyeris for message control

Introduces a 'verbose' parameter to plot.eyeris and related functions, allowing users to suppress status messages during plotting. Documentation and examples updated to reflect the new option.

shawntz committed:
Update bin.Rd

shawntz committed:
FF: #196 remove PDF report check from report rendering logic

The condition for rendering reports now only checks for html_report, not pdf_report, which was causing an error due to `pdf_report` being deprecated.

shawntz committed:
Update zzz.R

shawntz committed:
TEST: #18 add tests for `bin()` and `glassbox()` functions

Introduces a new test file to verify the behavior of the bin() function and its integration with glassbox(). Tests include checks for correct binning, sample rate, and absence of NA values, as well as ensuring no warnings are produced.

shawntz committed:
TEST: #18 add tests for `downsample()` and `glassbox()` functions

Introduces a new test file to verify the behavior of the downsample() function, including its integration with glassbox(). Tests ensure correct sample rate, row reduction, and absence of NA values after downsampling.

shawntz committed:
CHORE: #18 remove old downsample and bin combined unit tests

Deleted the test file for downsampling and binning functionality. This may be in preparation for refactoring, deprecation, or replacement of these tests.

shawntz committed:
CHORE: #18 add commented optional steps in data processing pipeline

Added commented lines for optional use of `downsample()` and `bin()` functions in the data processing pipeline, with notes to consult documentation before enabling. This provides users with more guidance on available processing options.

shawntz committed:
TEST: #206 add time monotonicity checks to pipeline functions

Introduces the check_time_monotonic utility to validate that time series columns are monotonically increasing. This check is now applied in bin_pupil and pipeline_handler for both single and block timeseries data, improving data integrity and error reporting for time-related operations.

Co-Authored-By: Mingjian He <49259768+mh105@users.noreply.github.com>

shawntz committed:
TEST: #206 add tests for time series monotonicity validation

Introduce unit tests to verify that the bin() function and check_time_monotonic() correctly validate monotonicity of time series data, including handling of non-monotonic, empty, single-point, and NA-containing sequences.

Co-Authored-By: Mingjian He <49259768+mh105@users.noreply.github.com>
